src/JsonReader.cs

```diff
@@ -476,42 +476,53 @@ sealed class NonProperty : IJsonProperty<Unit, JsonReadResult<Unit>>
                                                           Func<List<KeyValuePair<string, T>>, TResult> resultSelector) =>
         CreatePure((ref Utf8JsonReader rdr) =>
         {
-            if (rdr.TokenType is JsonTokenType.None && !rdr.Read())
-                throw PartialJsonNotSupportedException();
-
-            var tokenType = rdr.TokenType;
-            if (tokenType is not JsonTokenType.StartObject)
-                return Error("Invalid JSON value where a JSON object was expected.");
-
-            var properties = new List<KeyValuePair<string, T>>();
+            var (sm, currentPropertyName, acc) =
+                rdr.IsResuming
+                    ? ((ObjectReadStateMachine, string?, List<KeyValuePair<string, T>>))rdr.Pop()
+                    : (default, default, new());
 
             while (true)
             {
-                if (!rdr.TryReadToken(out tokenType))
-                    throw PartialJsonNotSupportedException();
-
-                if (tokenType is JsonTokenType.EndObject)
-                    break;
-
-                Debug.Assert(rdr.TokenType is JsonTokenType.PropertyName);
-                var name = rdr.GetString()!;
+                switch (sm.Read(ref rdr))
+                {
+                    case ObjectReadStateMachine.ReadResult.Error:
+                    {
+                        return Error("Invalid JSON value where a JSON object was expected.");
+                    }
+                    case ObjectReadStateMachine.ReadResult.Incomplete:
+                    {
+                        return rdr.Suspend((sm, currentPropertyName, acc));
+                    }
+                    case ObjectReadStateMachine.ReadResult.PropertyName:
+                    {
+                        currentPropertyName = rdr.GetString();
+                        sm.OnPropertyNameRead();
+                        break;
+                    }
+                    case ObjectReadStateMachine.ReadResult.PropertyValue:
+                    {
+                        switch (reader.TryRead(ref rdr))
+                        {
+                            case { Incomplete: true }:
+                                return rdr.Suspend((sm, currentPropertyName, acc));
+                            case { Error: { } err }:
+                                return new JsonReadError(err);
+                            case { Value: var val }:
+                                acc.Add(KeyValuePair.Create(currentPropertyName!, val));
```

Comment:
Prefer `Debug.Assert` over using `!`:

```suggestion
                                Debug.Assert(currentPropertyName is not null);
                                acc.Add(KeyValuePair.Create(currentPropertyName, val));
```